src/render/src/lib.rs

```diff
@@ -146,70 +140,259 @@ pub mod shade;
-            .collect::<Vec<_>>();
-        
+pub struct Backbuffer<B: Backend, Cf: RenderFormat> {
+    color: handle::RenderTargetView<B, Cf>,
+    // TODO: depth
+}
+
+use self::SyncState::*;
+#[derive(PartialEq)]
+enum SyncState {
+    Unreached,
+    Reached,
+}
+
+struct Sync<T> {
+    inner: T,
+    state: SyncState,
+}
+
+impl<T> Sync<T> {
+    fn reached(inner: T) -> Self {
+        Sync { inner, state: Reached }
+    }
+}
+
+struct FrameBundle<B: Backend> {
+    handles: handle::Bag<B>,
+    scopes: Vec<encoder::ScopeDependency<B, core::Graphics>>,
+    // wait until the backbuffer image is ready
+    wait_semaphore: Sync<B::Semaphore>,
+    // signal when the frame is done
+    signal_semaphore: Sync<B::Semaphore>,
+    signal_fence: Sync<B::Fence>,
+}
+
+impl<B: Backend> Context<B> {
+    pub fn init<Cf>(mut surface: B::Surface, adapter: B::Adapter)
+        -> (Self, Vec<Backbuffer<B, Cf>>)
+        where Cf: RenderFormat
+    {
         let core::Gpu {
-            device,
+            mut device,
             mut general_queues,
             mut graphics_queues,
             heap_types,
             memory_heaps,
             ..
-        } = adapter.open(&queue_descs);
+        } = adapter.open_with(|ref family, qtype| {
+            if qtype.supports_graphics() && surface.supports_queue(family) {
+                (1, QueueType::Graphics)
+            } else {
+                (0, QueueType::Transfer)
+            }
+        });
         
-        let queue = if general_queues.is_empty() {
-            Queue::Graphics(graphics_queues.remove(0))
-        } else {
-            Queue::General(general_queues.remove(0))
-        };
+        let queue = Queue::new(graphics_queues.remove(0));
+
+        let swap_config = core::SwapchainConfig::new()
+            .with_color::<Cf>();
+        // TODO: call supports_queue
+        let mut swapchain = surface.build_swapchain(swap_config, &queue.inner);
+        let backbuffers = Vec::<core::Backbuffer<B>>::new(); // TODO: take swapchain backbuffers
+
+        let frame_bundles = backbuffers.iter()
+            .map(|_| FrameBundle {
+                handles: handle::Bag::new(),
+                scopes: Vec::new(),
+                wait_semaphore: Sync::reached(device.create_semaphore()),
+                signal_semaphore: Sync::reached(device.create_semaphore()),
+                signal_fence: Sync::reached(device.create_fence(true)),
+            }).collect();
 
         let (garbage_sender, garbage_receiver) = handle::garbage_channel();
+        let mut device = Device::new(device, heap_types, memory_heaps, garbage_sender);
+
+        let backbuffers = backbuffers.into_iter()
+            .map(|raw| {
+                Backbuffer {
+                    color: Typed::new(device
+                        .view_backbuffer_as_render_target_raw(
+                            raw.color,
+                            unimplemented!(),
+                            unimplemented!(),
```

Comment:
How should I retrieve the backbuffer dimensions for `Kind` ? And what should I put in `Bind` ?

Reply:
Perhaps, we should add `Surface::get_kind()` method to the trait. For `Bind`, I think having `RENDER_TARGET` is required, and maybe we can have `TRANSFER_SRC`.